File: ClipboardZanager/Sources/ClipboardZanager.Core.Desktop/ComponentModel/SecurityHelper.cs
using System;
using System.Security;
using System.Security.Cryptography;
using System.Security.Permissions;
using System.Text;
using ClipboardZanager.Shared.Core;
using ClipboardZanager.Shared.Logs;

namespace ClipboardZanager.Core.Desktop.ComponentModel
{
    /// <summary>
    /// Provides a set of functions designed to manage the security
    /// </summary>
    internal static class SecurityHelper
    {
        #region Fields

        private static UIPermission _uiPermissionAllClipboard;
        private static SecurityPermission _unmanagedCodePermission;

        #endregion

        #region Methods

        /// <summary>
        /// Demands all the permission to access to all clipboards from all windows.
        /// </summary>
        [SecurityCritical]
        internal static void DemandAllClipboardPermission()
        {
            if (_uiPermissionAllClipboard == null)
            {
                _uiPermissionAllClipboard = new UIPermission(UIPermissionWindow.AllWindows, UIPermissionClipboard.AllClipboard);
            }
            _uiPermissionAllClipboard.Demand();
        }

        /// <summary>
        /// Demands the permission to work with unmanaged code.
        /// </summary>
        [SecurityCritical]
        internal static void DemandUnmanagedCode()
        {
            if (_unmanagedCodePermission == null)
            {
                _unmanagedCodePermission = new SecurityPermission(SecurityPermissionFlag.UnmanagedCode);
            }
            _unmanagedCodePermission.Demand();
            _unmanagedCodePermission.Assert();
        }

        /// <summary>
        /// Convert a <see cref="string"/> to a <see cref="SecureString"/>
        /// </summary>
        /// <param name="input">The string to secure</param>
        /// <returns>The <see cref="SecureString"/> that corresponds to the input string</returns>
        internal static SecureString ToSecureString(string input)
        {
            var secure = new SecureString();
            foreach (var c in input)
            {
                secure.AppendChar(c);
            }

            secure.MakeReadOnly();
            return secure;
        }

        /// <summary>
        /// Convert a <see cref="SecureString"/> to a <see cref="string"/>
        /// </summary>
        /// <param name="input">The secured string</param>
        /// <returns>Returns an unsecured string</returns>
        internal static string ToUnsecureString(SecureString input)
        {
            string returnValue;
            var ptr = System.Runtime.InteropServices.Marshal.SecureStringToBSTR(input);

            try
            {
                returnValue = System.Runtime.InteropServices.Marshal.PtrToStringBSTR(ptr);
            }
            finally
            {
                System.Runtime.InteropServices.Marshal.ZeroFreeBSTR(ptr);
            }

            return returnValue;
        }

        /// <summary>
        /// Encrypt a string
        /// </summary>
        /// <param name="input">The <see cref="SecureString"/> that contains the string's value</param>
        /// <returns>An encrypted <see cref="string"/></returns>
        internal static string EncryptString(SecureString input)
        {
            return EncryptString(input, ToSecureString(CoreHelper.GetApplicationVersion().ToString()));
        }

        /// <summary>
        /// Encrypt a string
        /// </summary>
        /// <param name="input">The <see cref="SecureString"/> that contains the string's value</param>
        /// <param name="password">The <see cref="SecureString"/> that represents the password</param>
        /// <returns>An encrypted <see cref="string"/></returns>
        internal static string EncryptString(SecureString input, SecureString password)
        {
            using (var rdb = GetSaltKeys(password))
            {
                var key = rdb.GetBytes(8);
                var iv = rdb.GetBytes(8);
                using (var algorithm = DES.Create())
                using (var transform = algorithm.CreateEncryptor(key, iv))
                {
                    var inputbuffer = Encoding.Unicode.GetBytes(ToUnsecureString(input));
                    var outputBuffer = transform.TransformFinalBlock(inputbuffer, 0, inputbuffer.Length);
                    return DataHelper.ToBase64(outputBuffer);
                }
            }
        }

        /// <summary>
        /// Decrypt an encrypted string
        /// </summary>
        /// <param name="encryptedData">The encrypted string</param>
        /// <returns>A <see cref="SecureString"/> that corresponds to the string's value</returns>
        internal static SecureString DecryptString(string encryptedData)
        {
            return DecryptString(encryptedData, ToSecureString(CoreHelper.GetApplicationVersion().ToString()));
        }

        /// <summary>
        /// Decrypt an encrypted string
        /// </summary>
        /// <param name="encryptedData">The encrypted string</param>
        /// <param name="password">The <see cref="SecureString"/> that represents the password</param>
        /// <returns>A <see cref="SecureString"/> that conrresponds to the string's value</returns>
        internal static SecureString DecryptString(string encryptedData, SecureString password)
        {
            try
            {
                using (var rdb = GetSaltKeys(password))
                {
                    var key = rdb.GetBytes(8);
                    var iv = rdb.GetBytes(8);

                    using (var algorithm = DES.Create())
                    using (var transform = algorithm.CreateDecryptor(key, iv))
                    {
                        var inputbuffer = DataHelper.ByteArrayFromBase64(encryptedData);
                        var outputBuffer = transform.TransformFinalBlock(inputbuffer, 0, inputbuffer.Length);
                        return ToSecureString(Encoding.Unicode.GetString(outputBuffer));
                    }
                }
            }
            catch (Exception ex)
            {
                Logger.Instance.Error(ex);
                return new SecureString();
            }
        }

        /// <summary>
        /// Generate a <see cref="Rfc2898DeriveBytes"/> that we can use to retrieve the KEY and IV.
        /// </summary>
        /// <param name="password">The password</param>
        /// <returns>A <see cref="Rfc2898DeriveBytes"/></returns>
        internal static Rfc2898DeriveBytes GetSaltKeys(SecureString password)
        {
            Requires.NotNull(password, nameof(password));
            return new Rfc2898DeriveBytes(ToUnsecureString(password), new byte[] { 0x53, 0x69, 0x75, 0x6f, 0x65, 0x20, 0x43, 0x69, 0x61, 0x68, 0x6d, 0x6c, 0x6f, 0x72, 0x64, 0x69, 0x64 });
        }

        #endregion
    }
}

File: Android/app/src/main/java/com/etiennebaudoux/clipboardzanager/componentmodel/core/SecurityHelper.java
package com.etiennebaudoux.clipboardzanager.componentmodel.core;

import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.KeySpec;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Provides a set of functions designed to manage the security
 */

public final class SecurityHelper {
    /**
     * Encrypt a {@link String}
     *
     * @param input The {@link String} to encrypt.
     * @return An encrypted {@link String}
     */
    public static String encryptString(String input) {
        return encryptString(input, CoreHelper.getApplicationVersion());
    }

    /**
     * Encrypt a {@link String}
     *
     * @param input    The {@link String} to encrypt.
     * @param password The {@link String} that represents the password
     * @return An encrypted {@link String}
     */
    public static String encryptString(String input, String password) {
        try {
            byte[] salt = getSaltKeys(password).getEncoded();
            byte[] key = new byte[8];
            byte[] iv = new byte[8];

            System.arraycopy(salt, 0, key, 0, 8);
            System.arraycopy(salt, 8, iv, 0, 8);
            Cipher cipher = Cipher.getInstance("DES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "DES"), new IvParameterSpec(iv));

            byte[] inputBuffer = input.getBytes(StandardCharsets.UTF_16LE);
            byte[] outputBuffer = cipher.doFinal(inputBuffer, 0, inputBuffer.length);
            return DataHelper.toBase64(outputBuffer);
        } catch (Exception ex) {
            return "";
        }
    }

    /**
     * Decrypt an encrypted {@link String}
     *
     * @param encryptedData The encrypted {@link String}
     * @return A {@link String} that corresponds to the {@link String}'s value
     */
    public static String decryptString(String encryptedData) {
        return decryptString(encryptedData, CoreHelper.getApplicationVersion());
    }

    /**
     * Decrypt an encrypted {@link String}
     *
     * @param encryptedData The encrypted {@link String}
     * @param password      The {@link String} that represents the password
     * @return A {@link String} that corresponds to the {@link String}'s value
     */
    public static String decryptString(String encryptedData, String password) {
        try {
            byte[] salt = getSaltKeys(password).getEncoded();
            byte[] key = new byte[8];
            byte[] iv = new byte[8];

            System.arraycopy(salt, 0, key, 0, 8);
            System.arraycopy(salt, 8, iv, 0, 8);
            Cipher cipher = Cipher.getInstance("DES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "DES"), new IvParameterSpec(iv));

            byte[] inputBuffer = DataHelper.byteArrayFromBase64(encryptedData);
            byte[] outputBuffer = cipher.doFinal(inputBuffer, 0, inputBuffer.length);
            return new String(outputBuffer, StandardCharsets.UTF_16LE);
        } catch (Exception ex) {
            return "";
        }
    }

    /**
     * Generate a {@link SecretKey} that we can use to retrieve the KEY and IV.
     *
     * @param password The password
     * @return a {@link SecretKey}
     * @throws NoSuchAlgorithmException
     * @throws InvalidKeySpecException
     */
    public static SecretKey getSaltKeys(String password) throws InvalidKeySpecException, NoSuchAlgorithmException {
        byte[] salt = new byte[]{0x53, 0x69, 0x75, 0x6f, 0x65, 0x20, 0x43, 0x69, 0x61, 0x68, 0x6d, 0x6c, 0x6f, 0x72, 0x64, 0x69, 0x64};
        return getSaltKeys(password, salt);
    }

    /**
     * Generate a {@link SecretKey} that we can use to retrieve the KEY and IV.
     *
     * @param password The password
     * @param salt     The salt key
     * @return a {@link SecretKey}
     * @throws NoSuchAlgorithmException
     * @throws InvalidKeySpecException
     */
    public static SecretKey getSaltKeys(String password, byte[] salt) throws NoSuchAlgorithmException, InvalidKeySpecException {
        Requires.notNullOrEmpty(password, "password");

        SecretKeyFactory secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
        KeySpec keySpec = new PBEKeySpec(password.toCharArray(), salt, 1000, 128);
        return secretKeyFactory.generateSecret(keySpec);
    }
}
